security/pkg/nodeagent/test/setup.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"google.golang.org/grpc"
+	ghc "google.golang.org/grpc/health/grpc_health_v1"
+
+	proxyEnv "istio.io/istio/mixer/test/client/env"
+	"istio.io/istio/pkg/spiffe"
+	istioEnv "istio.io/istio/pkg/test/env"
+	"istio.io/istio/security/pkg/nodeagent/cache"
+	citadel "istio.io/istio/security/pkg/nodeagent/caclient/providers/citadel"
+	"istio.io/istio/security/pkg/nodeagent/sds"
+	"istio.io/istio/security/pkg/nodeagent/secretfetcher"
+	caserver "istio.io/istio/security/pkg/nodeagent/test/mock"
+)
+
+const (
+	proxyTokenPath = "/tmp/sds-envoy-token.jwt"
+	sdsPath        = "/tmp/sdstestudspath"
+)
+
+// Env manages test setup and teardown.
+type Env struct {
+	ProxySetup           *proxyEnv.TestSetup
+	OutboundListenerPort int
+	// SDS server
+	SDSServer *sds.Server
+	// CA server
+	CAServer *caserver.CAServer
+}
+
+// TearDown tears down all components.
+func (e *Env) TearDown() {
+	// Stop proxy first, otherwise XDS stream is still alive and server's graceful
+	// stop will be blocked.
+	e.ProxySetup.TearDown()
+	e.SDSServer.Stop()
+	e.CAServer.GRPCServer.GracefulStop()
+}
+
+func getDataFromFile(filePath string, t *testing.T) []byte {
+	data, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		t.Fatalf("failed to read %q", filePath)
+	}
+	return data
+}
+
+// WriteDataToFile writes data into file
+func WriteDataToFile(path string, content []byte) error {
+	if path == "" {
+		return errors.New("empty file path")
+	}
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err = f.Write(content); err != nil {
+		return err
+	}
+	_ = f.Sync()
+	return nil
+}
+
+// SetupTest starts Envoy, SDS server, CA server and a dummy backend.
+// The test allow HTTP request flow.
+//                                 CA server
+//                                     |
+//                        +--------SDS server--------+
+//                        |                          |
+// HTTP    ->outbound ->outbound TLS <--mTLS-->inbound TLS ->inbound ->backend
+// request   listener   cluster                listener      cluster
+func SetupTest(t *testing.T, testID uint16) *Env {
+	// Set up credential files for bootstrap config
+	jwtToken := getDataFromFile(istioEnv.IstioSrc+"/security/pkg/nodeagent/test/testdata/jwt-token.jwt", t)
```

Comment:
nit: can you just use a string instead of loading from file?

Reply:
Done. Thanks.